rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,461 @@
+
+**Few points to note for React Native**
+
+- React Native [recently added](https://reactnative.dev/blog/2023/01/12/version-071) native support for `gap` specific properties in `v0.71`.
+- Polyfilling this on native for older releases seems bit hacky _(with cloning and wrapping child elements)_ and unneeded because unlike browsers, all users can receive an updated React Native v0.71 app, tradeoff being we will only be supporting `gap` based props in React Native going forward from `0.71`, so consumer apps should update their React Native version in order to consume these APIs.
+- No support for grid props since it isn't supported natively.
+- The TypeScript typings and autosuggestions should appropriately be adjusted for web and native to accommodate the differences, eg. grid properties, `inline-flex`, `block` display properties won't be supported. This can be done with [`moduleSuffixes`](https://www.typescriptlang.org/tsconfig#moduleSuffixes) detailed in the [related RFC](./writing-cross-platform-typescript.md).
+
+We have an [internal Box component](../packages/blade/src/components/Box/) but it exposes lot more properties than the ones associated with layout. It seems like a good idea to have a different public `Box` component with the restrictive but fully covered layout properties from above table.
+
+#### `Flex` component
+
+A helper component over the `Box` component with `display` set as `Flex`. Arguably, this component seems unneeded without flex specific shorthand properties because the behavior can be achieved with a `display="flex"` prop on `Box` but it improves code readability. Supports same properties.
+
+#### `Grid` component
+
+A helper **web only** component over the `Box` component with `display` set as `grid`. Similar to `Flex` component, it may be argued this component is unneeded with `display="grid"` set on `Box` but it improves code readability. Supports same properties.
+
+#### `Spacer` component
+
+A helper component over the `Box` component primarily aimed at putting space in between components. Useful when the spacing is different between each sibling and can't be solved by gap based properties. `Spacer` can also be helpful for spacing in between custom components which might not support a `sx` prop. Sample usage:
+
+```tsx
+<Box
+  display="flex"
+  flexDirection="column"
+  alignItems="center"
+  marginTop="spacing.1"
+  marginBottom="spacing.2"
+  paddingX="spacing.1"
+>
+  <Button>Button 1</Button>
+  <Spacer size="spacing.1" />
+  <Button>Button 2</Button>
+  <Spacer size="spacing.2" />
+  <Button>Button 3</Button>
+</Box>
+```
+
+Supported properties:
+
+| Prop     | Description                        |
+| -------- | ---------------------------------- |
+| `size`   | Sets both height and width equally |
+| `height` | Sets height                        |
+| `width`  | Sets width                         |
```

Comment:
Why do we need to add width & height support on spacer component? Shouldn't it instead expose margin & padding properties?

```jsx
<Spacer 
    margin={{ top: 'spacing.1', right: '3px', bottom: '5%', left: '2rem' }} 
    padding={{ top: 'spacing.1', right: '3px', bottom: '5%', left: '2rem' }}
```

Probably also along with responsive support

Reply:
Yup, these props should all support responsive properties. Regarding the prop names I feel `height` and `width` seems closer to the purpose of component.

Margin and padding seems like an overkill here because we're not rendering anything inside the spacer (just space).
- Should I use margin or padding for adding this space?
- If I need to add `spacing.2` space in between should I use `marginTop`, `marginBottom`, `paddingTop`, `paddingBottom` etc. Since I know the space I should be able to directly specify that with `size` or `height`
- Margin and padding are applied twice in a direction (top, bottom when vertically and left, right when horizontally) which will add another area to think about if applying spacing with margin padding properties

Reply:
Updated https://github.com/razorpay/blade/pull/950/commits/f4e3b6cebb79e8d1a3b72ed54eb53ed67c55dec4 to not expose `Spacer` from our side but in RFC have kept it as a recommendation which users may build over `Box` as a local abstraction.